agent/grpc-external/services/resource/read_test.go

```diff
@@ -213,42 +218,114 @@ func TestRead_VerifyReadConsistencyArg(t *testing.T) {
 // N.B. Uses key ACLs for now. See demo.RegisterTypes()
 func TestRead_ACLs(t *testing.T) {
 	type testCase struct {
-		authz resolver.Result
-		code  codes.Code
+		res          *pbresource.Resource
+		authz        resolver.Result
+		codeNotExist codes.Code
+		codeExists   codes.Code
 	}
+
+	artist, err := demo.GenerateV2Artist()
+	require.NoError(t, err)
+
+	label, err := demo.GenerateV1RecordLabel("blink1982")
+	require.NoError(t, err)
+
 	testcases := map[string]testCase{
-		"read hook denied": {
-			authz: AuthorizerFrom(t, demo.ArtistV1ReadPolicy),
-			code:  codes.PermissionDenied,
+		"artist-v1/read hook denied": {
+			res:          artist,
+			authz:        AuthorizerFrom(t, demo.ArtistV1ReadPolicy),
+			codeNotExist: codes.PermissionDenied,
+			codeExists:   codes.PermissionDenied,
+		},
+		"artist-v2/read hook allowed": {
+			res:          artist,
+			authz:        AuthorizerFrom(t, demo.ArtistV2ReadPolicy),
+			codeNotExist: codes.NotFound,
+			codeExists:   codes.OK,
 		},
-		"read hook allowed": {
-			authz: AuthorizerFrom(t, demo.ArtistV2ReadPolicy),
-			code:  codes.NotFound,
+		// Labels have the read ACL that requires reading the data.
+		"label-v1/read hook denied": {
+			res:          label,
+			authz:        AuthorizerFrom(t, demo.LabelV1ReadPolicy),
+			codeNotExist: codes.NotFound,
+			codeExists:   codes.PermissionDenied,
 		},
 	}
 
+	adminAuthz := AuthorizerFrom(t, `key_prefix "" { policy = "write" }`)
+
+	idx := 0
+	nextTokenContext := func(t *testing.T) context.Context {
+		// Each query should use a distinct token string to avoid caching so we can
+		// change the behavior each call.
+		token := fmt.Sprintf("token-%d", idx)
+		idx++
+		return context.WithValue(testContext(t), "x-consul-token", token)
+	}
+
 	for desc, tc := range testcases {
 		t.Run(desc, func(t *testing.T) {
 			server := testServer(t)
 			client := testClient(t, server)
 
-			mockACLResolver := &MockACLResolver{}
-			mockACLResolver.On("ResolveTokenAndDefaultMeta", mock.Anything, mock.Anything, mock.Anything).
-				Return(tc.authz, nil)
-			server.ACLResolver = mockACLResolver
-			demo.RegisterTypes(server.Registry)
+			dr := &dummyACLResolver{
+				result: testutils.ACLsDisabled(t),
+			}
+			server.ACLResolver = dr
 
-			artist, err := demo.GenerateV2Artist()
-			require.NoError(t, err)
+			demo.RegisterTypes(server.Registry)
 
-			// exercise ACL
-			_, err = client.Read(testContext(t), &pbresource.ReadRequest{Id: artist.Id})
-			require.Error(t, err)
-			require.Equal(t, tc.code.String(), status.Code(err).String())
+			dr.SetResult(tc.authz)
+			testutil.RunStep(t, "does not exist", func(t *testing.T) {
+				_, err = client.Read(nextTokenContext(t), &pbresource.ReadRequest{Id: tc.res.Id})
+				if tc.codeNotExist == codes.OK {
+					require.NoError(t, err)
+				} else {
+					require.Error(t, err)
+				}
+				require.Error(t, err)
+				require.Equal(t, tc.codeNotExist.String(), status.Code(err).String(), "%v", err)
+			})
```

Comment:
Is L286 necessary? It contradicts L282.

Reply:
In the "does not exist" path in the current form `codes.OK` isn't possible so it was still correct, but I've removed the stray line as it wasn't intentional.